File: bftoc.py
import sys

bf_file = open(sys.argv[1], "r").read()
c_output_file = open(sys.argv[2], "w")

ALLOWED_CHARS = ".,+-<>[]"
def del_unallowed_chars(string):
    for char in string:
        if char not in ALLOWED_CHARS:
            string = string.replace(char, "")
    return string

def bf_to_c(bf_file):
    bf_file = del_unallowed_chars(bf_file)
    c_output = ""
    is_in_loop = False
    for char in bf_file:
        c_output += "\t"
        if char == '>':
            c_output += "++p;\n"            
        elif char == '<':
            c_output += "--p;\n"
        elif char == '+':
            c_output += "++(*p);\n"
        elif char == '-':
            c_output += "--(*p);\n"
        elif char == '.':
            c_output += "putchar(*p);\n"
        elif char == ',':
            c_output += "*p = getchar();\n"
        elif char == '[':
            c_output += "while(*p) {\n"
            is_in_loop = True
        elif char == ']':
            c_output += "}\n"
            is_in_loop = False
        if is_in_loop:
            c_output += "\t"
    return c_output

c_compiled_program = bf_to_c(bf_file)

c_program_template = f"""\
#include <stdio.h>
int main() \x7b
    char memory[30000] = \x7b0\x7d;\n\
    char *p = memory;\n\
{c_compiled_program}
    return 0;
\x7d;"""

print(c_program_template, file=c_output_file)

File: bftoc.rs
use std::io::Write;

fn main() {
    let args: Vec<String> = std::env::args().collect();

    let in_file = std::fs::read(args[1].clone()).unwrap();
    let mut out_file = std::fs::File::create(args[2].clone()).unwrap();
    let data: String = del_unallowed_chars(in_file.iter().map(|&x| x as char).collect());
    
    let program = format!(r#"#include <stdio.h>
int main() {{
char memory[3000] = {{0}};
char *p = memory;
{}
return 0;
}};"#, bf_to_c(data));

    out_file.write_all(program.as_bytes()).unwrap();

}

fn del_unallowed_chars(data: String) -> String {
    let mut out = String::new();
    for c in data.chars() {
        if [".", ",", "+", "-", "<", ">", "[", "]"].contains(&c.to_string().as_str()) {
            out.push(c);
        }
    }
    out
}

fn bf_to_c(data: String) -> String {
    let mut out = String::new();
    for chr in data.chars() {
        match chr {
            '>' => out.extend("++p;\n".chars()),
            '<' => out.extend("--p;\n".chars()),
            '+' => out.extend("++(*p);\n".chars()),
            '-' => out.extend("--(*p);\n".chars()),
            '.' => out.extend("putchar(*p);\n".chars()),
            ',' => out.extend("*p = getchar();\n".chars()),
            '[' => out.extend("while(*p) {\n".chars()),
            ']' => out.extend("};\n".chars()),
            _ => {}
        }
    }
    out
}